sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,239 @@
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken) =>
+            await WaitForCompletionAsync(_defaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            _operation.AddAttributes(scope);
+            scope.Start();
+
+            try
+            {
+                Response response = async
+                    ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                    : _operation.GetResponse(cancellationToken);
+
+                RawResponse = response;
+
+                var status = _operation.UpdateState(response);
+
+                if (status == CompletionStatus.Succeeded)
+                {
+                    Value = _operation.ParseResponse(response);
+                    HasCompleted = true;
+                }
+                else if (status == CompletionStatus.Failed)
+                {
+                    OperationFailedException = _operation.GetOperationFailedException(response);
+                    HasCompleted = true;
+
+                    throw OperationFailedException;
+                }
+
+                return response;
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+    }
+
+#pragma warning disable SA1402 // File may only contain a single type
+    internal class OperationInternal<TResult, TResponseType> : OperationInternal<TResult>
+#pragma warning restore SA1402 // File may only contain a single type
```

Comment:
Put everything in the same file for simplicity, but we can separate stuff if we think it's better. We would end up with 5 files, though (two classes, two interfaces, one enum).